File: VezioCastelManager/Backend/veziocastelmanager/src/main/java/com/generation/veziocastelmanager/service/VisitorService.java
package com.generation.veziocastelmanager.service;

import com.generation.veziocastelmanager.dto.VisitorDTO;
import com.generation.veziocastelmanager.mapper.VisitorMapper;
import com.generation.veziocastelmanager.model.entities.Visitor;
import com.generation.veziocastelmanager.model.repository.VisitorRepository;
import jakarta.persistence.EntityNotFoundException;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.validation.annotation.Validated;

import java.util.List;

@Service
@Validated
public class VisitorService
{
    @Autowired
    private VisitorRepository   visitorRepository;

    @Autowired
    private VisitorMapper       visitorMapper;

    // usato da GET /vcm/api/visitors
    // prendo tutti i visitatori dal db e li converto in DTO
    public List<VisitorDTO> findAll()
    {
        return visitorMapper.toDTOs(visitorRepository.findAll());
    }

    // usato da GET /vcm/api/visitors/{id}
    // cerco il visitatore per id, se non esiste lancio un'eccezione che Spring traduce in 404
    public VisitorDTO findById(int id)
    {
        Visitor visitor = visitorRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Visitor not found with id: " + id));
        return (visitorMapper.toDTO(visitor));
    }

    // usato da POST /vcm/api/visitors e PUT /vcm/api/visitors/{id}
    // @Valid fa scattare le validazioni definite nel DTO prima di procedere
    // il mapper converte il DTO in entità, salvo e riconverto in DTO per la risposta
    public VisitorDTO       save(@Valid VisitorDTO visitorDTO)
    {
        Visitor visitor = visitorMapper.toEntity(visitorDTO);
        visitor         = visitorRepository.save(visitor);
        return (visitorMapper.toDTO(visitor));
    }

    // usato da DELETE /vcm/api/visitors/{id}
    public void deleteById(int id)
    {
        visitorRepository.deleteById(id);
    }
}

File: VezioCastelManager/Frontend/src/app/services/visitor-service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Visitor } from '../model/entities';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root',
})
export class VisitorService
{
  private url = 'http://localhost:8080/vcm/api/visitors';
  http = inject(HttpClient);

  // GET /visitors
  findAll(): Observable<Visitor[]>
  {
    const request = this.http.get<Visitor[]>(this.url);
    return request;
  }

  // POST /visitors — crea un nuovo visitatore
  save(visitor: Visitor): Observable<Visitor>
  {
    const request = this.http.post<Visitor>(this.url, visitor);
    return request;
  }
}
